frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -941,106 +938,63 @@ function ReaderBookmark:isBookmarkAdded(item)
 end
 
 function ReaderBookmark:removeHighlight(item)
+    self:removeBookmark(item)
     if item.pos0 then
         self.ui:handleEvent(Event:new("Unhighlight", item))
-    else
-        self:removeBookmark(item)
-        -- Update dogear in case we removed a bookmark for current page
-        if self.ui.document.info.has_pages then
-            self:setDogearVisibility(self.view.state.page)
-        else
-            self:setDogearVisibility(self.ui.document:getXPointer())
-        end
+    else -- dogear bookmark, update it in case we removed a bookmark for current page
+        self:setDogearVisibility(self:getCurrentPageNumber())
     end
 end
 
--- binary search to remove bookmark
 function ReaderBookmark:removeBookmark(item, reset_auto_text_only)
-    local _middle
-    local _start, _end = 1, #self.bookmarks
-    while _start <= _end do
-        _middle = math.floor((_start + _end)/2)
-        local v = self.bookmarks[_middle]
-        if item.datetime == v.datetime and item.page == v.page then
-            if reset_auto_text_only then
-                if self:isBookmarkAutoText(v) then
-                    v.text = nil
-                end
-            else
-                self.ui:handleEvent(Event:new("BookmarkRemoved", v))
-                table.remove(self.bookmarks, _middle)
-                self.view.footer:onUpdateFooter(self.view.footer_visible)
-            end
-            return
-        elseif self:isBookmarkInPositionOrder(item, v) then
-            _end = _middle - 1
-        else
-            _start = _middle + 1
-        end
-    end
-    -- If we haven't found item, it may be because there are multiple
+    -- If we haven't found item in binary search, it may be because there are multiple
     -- bookmarks on the same page, and the above binary search decided to
     -- not search on one side of one it found on page, where item could be.
     -- Fallback to do a full scan.
-    logger.dbg("removeBookmark: binary search didn't find bookmark, doing full scan")
-    for i=1, #self.bookmarks do
-        local v = self.bookmarks[i]
-        if item.datetime == v.datetime and item.page == v.page then
-            if reset_auto_text_only then
-                if self:isBookmarkAutoText(v) then
-                    v.text = nil
-                end
-            else
-                self.ui:handleEvent(Event:new("BookmarkRemoved", v))
-                table.remove(self.bookmarks, i)
-                self.view.footer:onUpdateFooter(self.view.footer_visible)
-            end
-            return
+    local index = self:getBookmarkIndexBinary(item) or self:getBookmarkIndex(item)
```

Comment:
The naming of the first method is a bit bad - we're not getting a bookmark index "binary" :)
`getBookmarkIndexUsingBinarySearch(item) or self:getBookmarkIndexUsingFullScan(item)`
or
`getFastBookmarkIndex(item) or self:getFallbackBookmarkIndex(item)` (so-so)
or something better than that :)

Btw, you told me before, but why do we need these 2 ways of getting a bookmark index?

Reply:
> why do we need these 2 ways of getting a bookmark index?

Non-binary search is used when changing highlight boundaries (pos0 and pos1).
We can avoid this by passing the old highlight to find the corresponding bookmark.

And as a fallback
https://github.com/koreader/koreader/blob/5696a7cf00e2c8b6dd5d01adc9e6601252cd876b/frontend/apps/reader/modules/readerbookmark.lua#L981-L984
I did not investigate if there were any fails of binary search.